File: installer/installer_utils/platforms/linux.py
from .base import PlatformHandler
from ..config import InstallConfig
from ..utils import check_dependency_installed, run_command
import os


class LinuxHandler(PlatformHandler):
    def install_dependencies(self, config: InstallConfig):
        missing_dependencies = [
            cmd
            for cmd in ["git", "python3"]
            if not check_dependency_installed(f"which {cmd}")
        ]

        if missing_dependencies:
            raise RuntimeError(
                f"Missing required dependencies: {', '.join(missing_dependencies)}"
            )

        if config.timesync:
            run_command("timedatectl set-ntp true")

    def setup_autostart(self, config: InstallConfig):
        autostart_dir = os.path.expanduser("~/.config/autostart")
        os.makedirs(autostart_dir, exist_ok=True)
        desktop_file = os.path.join(autostart_dir, "octane.desktop")
        with open(desktop_file, "w") as f:
            f.write(
                "[Desktop Entry]\n"
                "Type=Application\n"
                f"Exec={os.path.join(config.install_path, '.venv/bin/python')} "
                f"{os.path.join(config.install_path, 'main.py')}\n"
                "Hidden=false\n"
                "NoDisplay=false\n"
                "X-GNOME-Autostart-enabled=true\n"
                "Name=Octane\n"
            )



    def get_activate_cmd(self, config: InstallConfig):
        return f"source {config.install_path}/.venv/bin/activate"

    def get_create_env_cmd(self, config: InstallConfig):
        return f"python3 -m venv {config.install_path}/.venv"

    def get_python_cmd(self):
        return "python3"

    def get_env_file(self):
        return ".env-example-unix"

File: source/Octane-Manager/installer_utils/platforms_specific/linux.ts
import * as fs from 'fs';
import * as path from 'path';
import { EnvironmentUtil } from '../modules/environment_util';
import { InstallerBase } from '../modules/installer_base';
import { DependencyInstaller } from '../modules/dependency_installer';
import { InstallConfig } from '../config';
import { check_dependency_installed, run_command } from '../utils';

class LinuxInstallerBase extends InstallerBase {
    setup_autostart(config: InstallConfig): void {
        const autostart_dir = path.join(os.homedir(), ".config", "autostart");
        fs.mkdirSync(autostart_dir, { recursive: true });
        const desktop_file = path.join(autostart_dir, "octane.desktop");
        const desktop_content = `
[Desktop Entry]
Type=Application
Exec=${path.join(config.install_path, '.venv/bin/python')} ${path.join(config.install_path, 'main.py')}
Hidden=false
NoDisplay=false
X-GNOME-Autostart-enabled=true
Name=Octane
        `;
        fs.writeFileSync(desktop_file, desktop_content);
    }
}

class LinuxDependencyInstaller extends DependencyInstaller {
    install_system_dependencies(config: InstallConfig): void {
        const missing_dependencies = ["git", "python3"].filter(cmd => !check_dependency_installed(`which ${cmd}`));

        if (missing_dependencies.length > 0) {
            throw new Error(`Missing required dependencies: ${missing_dependencies.join(', ')}`);
        }

        if (config.timesync) {
            run_command("timedatectl set-ntp true");
        }
    }
}

class LinuxEnvironmentUtil extends EnvironmentUtil {
    // Additional Linux-specific environment setup can be added here
}
